packages/tools/src/tools/displayTools/Labelmap/labelmapDisplay.ts

```diff
@@ -192,30 +210,55 @@ async function render(
   const segmentation = SegmentationState.getSegmentation(segmentationId);
   const labelmapData =
     segmentation.representationData[Representations.Labelmap];
-  const { volumeId: labelmapUID } = labelmapData;
 
-  const labelmap = cache.getVolume(labelmapUID);
+  let actorEntry;
+  if (labelmapData.type === 'volume') {
+    const { volumeId: labelmapUID } = labelmapData;
 
-  if (!labelmap) {
-    throw new Error(`No Labelmap found for volumeId: ${labelmapUID}`);
-  }
+    const labelmap = cache.getVolume(labelmapUID);
 
-  if (!isSameFrameOfReference(viewport, labelmapData?.referencedVolumeId)) {
-    return;
-  }
-  let actorEntry = viewport.getActor(segmentationRepresentationUID);
+    if (!labelmap) {
+      throw new Error(`No Labelmap found for volumeId: ${labelmapUID}`);
+    }
 
-  if (!actorEntry) {
-    const segmentation = SegmentationState.getSegmentation(segmentationId);
-    const { volumeId } =
-      segmentation.representationData[Representations.Labelmap];
-    // only add the labelmap to ToolGroup viewports if it is not already added
-    await _addLabelmapToViewport(
-      viewport,
-      volumeId,
-      segmentationRepresentationUID
-    );
+    if (!isSameFrameOfReference(viewport, labelmapData?.referencedVolumeId)) {
+      return;
+    }
+    actorEntry = viewport.getActor(segmentationRepresentationUID);
+
+    if (!actorEntry) {
+      // only add the labelmap to ToolGroup viewports if it is not already added
+      await _addLabelmapToViewport(
+        viewport,
+        labelmapData,
+        segmentationRepresentationUID
+      );
+    }
+
+    actorEntry = viewport.getActor(segmentationRepresentationUID);
+  } else {
+    const imageId = viewport.getCurrentImageId();
+    const { referencedImageIds } = labelmapData;
+    if (referencedImageIds.includes(imageId)) {
+      actorEntry = viewport.getActor(segmentationRepresentationUID);
+      if (!actorEntry) {
+        // only add the labelmap to ToolGroup viewports if it is not already added
+        await _addLabelmapToViewport(
+          viewport,
+          labelmapData,
+          segmentationRepresentationUID
+        );
+      }
 
+      viewport.element.removeEventListener(
+        Enums.Events.IMAGE_RENDERED,
+        updateSegmentationImage
+      );
+      viewport.element.addEventListener(
+        Enums.Events.IMAGE_RENDERED,
+        updateSegmentationImage
+      );
```

Comment:
These should go inside tools/src/eventListeners. You need a new StackImageChangeEventListener that gets initiated on the init and triggers `triggerSegmentationRender` or similar to make the flow reach here. You shouldn't subscribe here.